File: src/main/java/com/example/librarymanagement/repository/ReservationRepository.java
package com.example.librarymanagement.repository;

import com.example.librarymanagement.model.entity.Reservation;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Repository;

import java.util.List;

@Repository
public interface ReservationRepository extends JpaRepository<Reservation, Long> {
    List<Reservation> findAllByBookIdAndUserId(Long bookId, Long userId);
    List<Reservation> findAllByLibraryId(Long libraryId);
    List<Reservation> findAllByUserId(Long userId);
    Reservation findFirstByBookIdAndLibraryIdOrLibraryIsNull(Long bookId, Long libraryId);
}

File: src/main/java/com/example/librarymanagement/repository/ReservationRepository.kt
package com.example.librarymanagement.repository

import com.example.librarymanagement.model.domain.Reservation

interface ReservationRepository {
    fun save(reservation: Reservation): Reservation
    fun findById(reservationId: String): Reservation?
    fun deleteById(reservationId: String)
    fun findAllByBookIdAndUserId(bookId: String, userId: String): List<Reservation>
    fun findAllByLibraryId(libraryId: String): List<Reservation>
    fun findAllByUserId(userId: String): List<Reservation>
    fun findAllByBookId(bookId: String): List<Reservation>
    fun findFirstByBookIdAndLibraryIdOrLibraryIsNull(bookId: String, libraryId: String?): Reservation?
    fun existsByBookIdAndUserId(bookId: String, userId: String): Boolean
    fun findAllByBookIdAndLibraryId(bookId: String, libraryId: String): List<Reservation>
}
